File: email-worker/metrics/metrics.go
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// Email processing metrics
	EmailJobsProcessed = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "email_jobs_processed_total",
			Help: "Total number of email jobs processed",
		},
		[]string{"status", "job_type", "template"},
	)

	EmailJobProcessingDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "email_job_processing_duration_seconds",
			Help:    "Time spent processing email jobs",
			Buckets: []float64{0.1, 0.5, 1, 2, 5, 10, 30},
		},
		[]string{"job_type", "template"},
	)

	EmailQueueSize = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "email_queue_size",
			Help: "Current number of emails in queue",
		},
	)

	EmailDeliverySuccess = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "email_delivery_success_total",
			Help: "Total number of successful email deliveries",
		},
		[]string{"provider", "template"},
	)

	EmailDeliveryFailure = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "email_delivery_failure_total",
			Help: "Total number of failed email deliveries",
		},
		[]string{"provider", "template", "error_type"},
	)

	// HTTP metrics
	HttpRequestsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_total",
			Help: "Total number of HTTP requests",
		},
		[]string{"method", "endpoint", "status_code"},
	)

	HttpRequestDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "http_request_duration_seconds",
			Help:    "Duration of HTTP requests in seconds",
			Buckets: []float64{0.1, 0.3, 0.5, 0.7, 1, 3, 5, 7, 10},
		},
		[]string{"method", "endpoint"},
	)

	// gRPC metrics
	GrpcRequestsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "grpc_requests_total",
			Help: "Total number of gRPC requests",
		},
		[]string{"method", "status"},
	)

	GrpcRequestDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "grpc_request_duration_seconds",
			Help:    "Duration of gRPC requests in seconds",
			Buckets: []float64{0.01, 0.05, 0.1, 0.5, 1, 2, 5},
		},
		[]string{"method"},
	)

	// Database metrics
	DbQueryDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "db_query_duration_seconds",
			Help:    "Duration of database queries in seconds",
			Buckets: []float64{0.001, 0.01, 0.1, 0.5, 1, 2, 5},
		},
		[]string{"operation", "table"},
	)

	DbConnectionsActive = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "db_connections_active",
			Help: "Number of active database connections",
		},
	)

	// Cache metrics
	CacheHits = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "cache_hits_total",
			Help: "Total number of cache hits",
		},
		[]string{"cache_type"},
	)

	CacheMisses = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "cache_misses_total",
			Help: "Total number of cache misses",
		},
		[]string{"cache_type"},
	)

	// Business metrics
	BusinessEventsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "business_events_total",
			Help: "Total number of business events",
		},
		[]string{"event_type"},
	)
)

func Init() {
	// Register all metrics
	prometheus.MustRegister(EmailJobsProcessed)
	prometheus.MustRegister(EmailJobProcessingDuration)
	prometheus.MustRegister(EmailQueueSize)
	prometheus.MustRegister(EmailDeliverySuccess)
	prometheus.MustRegister(EmailDeliveryFailure)
	prometheus.MustRegister(HttpRequestsTotal)
	prometheus.MustRegister(HttpRequestDuration)
	prometheus.MustRegister(GrpcRequestsTotal)
	prometheus.MustRegister(GrpcRequestDuration)
	prometheus.MustRegister(DbQueryDuration)
	prometheus.MustRegister(DbConnectionsActive)
	prometheus.MustRegister(CacheHits)
	prometheus.MustRegister(CacheMisses)
	prometheus.MustRegister(BusinessEventsTotal)

	// Initialize auto metrics
	promauto.NewGaugeFunc(prometheus.GaugeOpts{
		Name: "email_worker_info",
		Help: "Information about the email worker",
	}, func() float64 {
		return 1
	})
} 

File: shared-lib/metrics.js
import prometheus from "prom-client";

// Common HTTP metrics
export const httpRequestDuration = new prometheus.Histogram({
  name: "http_request_duration_seconds",
  help: "Duration of HTTP requests in seconds",
  labelNames: ["method", "route", "status_code", "service"],
  buckets: [0.1, 0.3, 0.5, 0.7, 1, 3, 5, 7, 10],
});

export const httpRequestsTotal = new prometheus.Counter({
  name: "http_requests_total",
  help: "Total number of HTTP requests",
  labelNames: ["method", "route", "status_code", "service"],
});

export const httpRequestSize = new prometheus.Histogram({
  name: "http_request_size_bytes",
  help: "Size of HTTP requests in bytes",
  labelNames: ["method", "route", "service"],
  buckets: [100, 1000, 5000, 10000, 50000, 100000],
});

export const httpResponseSize = new prometheus.Histogram({
  name: "http_response_size_bytes",
  help: "Size of HTTP responses in bytes",
  labelNames: ["method", "route", "status_code", "service"],
  buckets: [100, 1000, 5000, 10000, 50000, 100000],
});

// gRPC metrics
export const grpcRequestDuration = new prometheus.Histogram({
  name: "grpc_request_duration_seconds",
  help: "Duration of gRPC requests in seconds",
  labelNames: ["method", "service", "status"],
  buckets: [0.01, 0.05, 0.1, 0.5, 1, 2, 5],
});

export const grpcRequestsTotal = new prometheus.Counter({
  name: "grpc_requests_total",
  help: "Total number of gRPC requests",
  labelNames: ["method", "service", "status"],
});

// Database metrics
export const dbQueryDuration = new prometheus.Histogram({
  name: "db_query_duration_seconds",
  help: "Duration of database queries in seconds",
  labelNames: ["operation", "table", "service"],
  buckets: [0.001, 0.01, 0.1, 0.5, 1, 2, 5],
});

export const dbConnectionsActive = new prometheus.Gauge({
  name: "db_connections_active",
  help: "Number of active database connections",
  labelNames: ["service"],
});

// Cache metrics
export const cacheHits = new prometheus.Counter({
  name: "cache_hits_total",
  help: "Total number of cache hits",
  labelNames: ["cache_type", "service"],
});

export const cacheMisses = new prometheus.Counter({
  name: "cache_misses_total",
  help: "Total number of cache misses",
  labelNames: ["cache_type", "service"],
});

// Business metrics
export const businessEventsTotal = new prometheus.Counter({
  name: "business_events_total",
  help: "Total number of business events",
  labelNames: ["event_type", "service"],
});

// Service-specific metrics
export const serviceMetrics = {
  gateway: {
    activeConnections: new prometheus.Gauge({
      name: "gateway_active_connections",
      help: "Number of active connections to gateway",
    }),
    upstreamLatency: new prometheus.Histogram({
      name: "gateway_upstream_latency_seconds",
      help: "Upstream service latency",
      labelNames: ["upstream_service"],
      buckets: [0.01, 0.05, 0.1, 0.5, 1, 2, 5],
    }),
    rateLimitExceeded: new prometheus.Counter({
      name: "gateway_rate_limit_exceeded_total",
      help: "Total number of rate limit violations",
      labelNames: ["client_ip", "endpoint"],
    }),
  },
  auth: {
    authAttempts: new prometheus.Counter({
      name: "auth_attempts_total",
      help: "Total authentication attempts",
      labelNames: ["method", "status", "user_type"],
    }),
    activeSessions: new prometheus.Gauge({
      name: "auth_active_sessions",
      help: "Number of active sessions",
    }),
    jwtValidations: new prometheus.Counter({
      name: "jwt_validations_total",
      help: "Total JWT token validations",
      labelNames: ["status"],
    }),
    passwordResetRequests: new prometheus.Counter({
      name: "password_reset_requests_total",
      help: "Total password reset requests",
      labelNames: ["status"],
    }),
  },
  email: {
    emailsSent: new prometheus.Counter({
      name: "emails_sent_total",
      help: "Total emails sent",
      labelNames: ["template", "status", "provider"],
    }),
    emailQueueSize: new prometheus.Gauge({
      name: "email_queue_size",
      help: "Current email queue size",
    }),
    emailProcessingDuration: new prometheus.Histogram({
      name: "email_processing_duration_seconds",
      help: "Email processing duration in seconds",
      labelNames: ["template"],
      buckets: [0.1, 0.5, 1, 2, 5, 10],
    }),
  },
};

// Metrics middleware for Express
export const metricsMiddleware = (serviceName) => (req, res, next) => {
  const start = Date.now();
  const originalSend = res.send;

  // Track request size
  const requestSize = req.headers["content-length"] || 0;
  httpRequestSize
    .labels(req.method, req.route?.path || req.path, serviceName)
    .observe(parseInt(requestSize));

  // Override send to track response size
  res.send = function (data) {
    const responseSize = Buffer.byteLength(data);
    httpResponseSize
      .labels(
        req.method,
        req.route?.path || req.path,
        res.statusCode,
        serviceName
      )
      .observe(responseSize);

    originalSend.call(this, data);
  };

  res.on("finish", () => {
    const duration = (Date.now() - start) / 1000;

    httpRequestDuration
      .labels(
        req.method,
        req.route?.path || req.path,
        res.statusCode,
        serviceName
      )
      .observe(duration);

    httpRequestsTotal
      .labels(
        req.method,
        req.route?.path || req.path,
        res.statusCode,
        serviceName
      )
      .inc();
  });

  next();
};

// gRPC metrics interceptor
export const grpcMetricsInterceptor = (serviceName) => {
  return {
    intercept: (call, nextCall) => {
      const start = Date.now();

      return new prometheus.Gauge({
        name: "grpc_active_requests",
        help: "Number of active gRPC requests",
        labelNames: ["method", "service"],
      }).inc();

      return nextCall(call).then(
        (response) => {
          const duration = (Date.now() - start) / 1000;
          grpcRequestDuration
            .labels(call.method, serviceName, "OK")
            .observe(duration);
          grpcRequestsTotal.labels(call.method, serviceName, "OK").inc();
          return response;
        },
        (error) => {
          const duration = (Date.now() - start) / 1000;
          grpcRequestDuration
            .labels(call.method, serviceName, "ERROR")
            .observe(duration);
          grpcRequestsTotal.labels(call.method, serviceName, "ERROR").inc();
          throw error;
        }
      );
    },
  };
};

// Metrics endpoint
export const metricsEndpoint = async (req, res) => {
  res.set("Content-Type", prometheus.register.contentType);
  res.end(await prometheus.register.metrics());
};

// Health check endpoint
export const healthCheckEndpoint = (req, res) => {
  res.json({
    status: "healthy",
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
  });
};

// Initialize default metrics
prometheus.collectDefaultMetrics({
  prefix: "booking_system_",
  labels: { service: "unknown" },
});

export default {
  metricsMiddleware,
  grpcMetricsInterceptor,
  metricsEndpoint,
  healthCheckEndpoint,
  serviceMetrics,
  prometheus,
};
